keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,819 @@
+* Expand the admission surface area (admission.Attributes, AdmissionReview, CEL authorizer implementation available to admission) 
+  since admission verbs don't support field/label selectors
+
+
+## Proposal
+
+List, Watch, and DeleteCollection requests directly have field and label selector options.
+A single-item List or Watch request is still a list as normal (including selectors), but also includes a name.
+
+### Authorization Attributes changes
+
+The authorization attributes have easy access to the query parameter field and label selectors.
+To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
+selector has no semantic meaning.
+In practice this means that (for now), only List, Watch, and DeleteCollection have field and label selectors.
+SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and
+label selector attributes removed prior to checking authorization.
+Webhook authors: remember that the list of verbs accepting field and label selectors may change over time.
+If the kube-apiserver sends the FieldSelector or LabelSelector to a webhook, the kube-apiserver intends to honor the selector attributes.
+Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize without
+field or label selectors since that will authorize using a wider permission (field and label selectors can only reduce access).
+
+```go
+type Attributes interface {
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+	// It can return an error if the field selector cannot be parsed.
+    // Remember that field selector formats vary based on the version of the API being used!
+	ParseFieldSelector() ([]Requirement, error)
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+#### Future-proofing your authentication webhook for future verbs
+
+As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
+<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
+In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.
+<<[/UNRESOLVED]>>
+* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
+* For Update, this means that the finalNewObject and oldObject must match the field and label selector.
+* For Delete, this means that the oldObject must match the field and label selector.
+ 
+We do not expand field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
+field selector and use a List to get equivalent functionality.
```

Comment:
@jpbetz @sttts **If** we add field and label selectors, I'd like to provide advice to the authentication webhook authors about what the semantics will be.  This is not a proposal to make those changes.

My intent is that if a field and/or label selector are specified on a mutation request, the object being mutated (both old and new) are guaranteed to match the selectors.  If the object (old or new) does not match the selector, the storage layer rejects the request.

Reply:
Do we need to define meaning of selectors if/when included for subresources in the future?

The main question I'd have is whether a distinction is made for subresources which do and do not see the full object in question (admission defines `objectSelector` to only work for subresources which see the full object)

* examples of ones which see the full object: `pods/status`
* examples of ones which do not see the full object: `pods/{exec,attach,log,proxy,binding}`, `deployments/scale`

Reply:
>The main question I'd have is whether a distinction is made for subresources which do and do not see the full object in question (admission defines objectSelector to only work for subresources which see the full object)

I'm nearly in agreement.  But I think that the storage layer of deployments/scale has the entire object. Versus admission which does not have the entire object.  `store *genericregistry.Store` appears to be shared between spec, status, and scale.

Subresources that are not stored into etcd would not be able to enforce as uniformally.  We could select specific subresources that we know end up in storage like status and scale.  pods/binding and eviction would both be useful, but far narrower.

Reply:
> if a field and/or label selector are specified on a mutation request, the object being mutated (both old and new) are guaranteed to match the selectors

I think this generalization works for subresources as well.

The storage layer must ensure the parent object matches the selectors and that the request does not change that (either because the request cannot modify labels / fields or after applying the mutations in the request it still matches the selectors).

If the storage layer cannot verify the parent object matches the selectors because it doesn't have visibility, the request would be rejected.

(for interested bystanders, this is all *very* hypothetical future oriented ... there is no current plan to add selectors to mutating requests or to subresources ... we just want the meaning to be well-defined so an authorizer making decisions today knows what it could be allowing in the future)